gui/widgets/grid.cpp

```diff
@@ -0,0 +1,640 @@
+/* ScummVM - Graphic Adventure Engine
+ *
+ * ScummVM is the legal property of its developers, whose names
+ * are too numerous to list here. Please refer to the COPYRIGHT
+ * file distributed with this source distribution.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#include "common/system.h"
+#include "common/file.h"
+#include "common/language.h"
+#include "common/platform.h"
+
+#include "gui/gui-manager.h"
+#include "gui/widgets/grid.h"
+
+#include "gui/ThemeEval.h"
+
+namespace GUI {
+
+GridItemWidget::GridItemWidget(GridWidget *boss, int x, int y, int w, int h)
+	: ContainerWidget(boss, x, y, w, h), CommandSender(boss) {
+
+	setFlags(WIDGET_ENABLED | WIDGET_TRACK_MOUSE | WIDGET_CLEARBG);
+	_activeEntry = nullptr;
+	_grid = boss;
+	isHighlighted = false;
+}
+GridItemWidget::GridItemWidget(GridWidget *boss)
+	: GridItemWidget(boss, 0, 0, 0, 0) {}
+
+void GridItemWidget::setActiveEntry(GridItemInfo &entry) {
+	_activeEntry = &entry;
+}
+
+void GridItemWidget::updateThumb() {
+	const Graphics::ManagedSurface *gfx = _grid->filenameToSurface(_activeEntry->thumbPath);
+	_thumbGfx.free();
+	if (gfx)
+		_thumbGfx.copyFrom(*gfx);
+}
+
+void GridItemWidget::update() {
+	if (_activeEntry) {
+		updateThumb();
+		setTooltip(_activeEntry->title);
+		markAsDirty();
+	}
+}
+
+void GridItemWidget::move(int x, int y) {
+	Widget::setPos(getRelX() + x, getRelY() + y);
+}
+
+void GridItemWidget::drawWidget() {
+	int thumbHeight = _grid->getThumbnailHeight();
+	int thumbWidth = _grid->getThumbnailWidth();
+	Array<U32String> titleLines;
+	g_gui.getFont().wordWrapText(_activeEntry->title, thumbWidth, titleLines);
+
+	if ((isHighlighted) || (_grid->getSelected() == _activeEntry->entryID)) {
+		// Draw a highlighted BG on hover
+		g_gui.theme()->drawWidgetBackground(Common::Rect(_x - (_grid->_gridXSpacing / 3), _y - (_grid->_gridYSpacing / 3),
```

Comment:
Consider declaring the rect on a line above, to make this slightly less wide, and more readable.

Reply:
3? Consider creating a constant to help explain the constant.

Reply:
Also, consider precalculating
_grid->_gridXSpacing / 3
_grid->_gridYSpacing / 3

Into some variables above the if-else, since they're used twice in each branch. Could help with the width a bit a s well.

Reply:
I have defined the Rect beforehand to save on width, I have added the X and Y margins as gridSpacing / 3 as constants to reduce recalculations in 24c9899, and I have added comments explaining the reason for 1/3 in 84ec5b6.